From: ohering@suse.de
Subject: netback: fix refounting
Patch-mainline: Never, SUSE-Xen specific
References: bsc#978094

netif_disconnect waits for the refcnt to drop. This might be done in the
netbk/N thread via netif_put. If the refcnt reaches zero wake_up is
called, which triggers a recheck of the wait_event condition. Since the
refcnt just reached zero the condition remains false and xenwatch waits
forever. If the refcnt starts at 1 then wake_up is never called.

--- a/drivers/xen/netback/interface.c
+++ b/drivers/xen/netback/interface.c
@@ -327,7 +327,8 @@ void netif_disconnect(struct backend_inf
 		netif_put(netif);
 	}
 
-	wait_event(netif->waiting_to_free, atomic_read(&netif->refcnt) == 1);
+	atomic_dec(&netif->refcnt);
+	wait_event(netif->waiting_to_free, atomic_read(&netif->refcnt) == 0);
 
 	del_timer_sync(&netif->credit_timeout);
 	del_timer_sync(&netif->tx_queue_timeout);
